inconsistent batch count in `pebble sync`

```
$ pebble sync -c 24 -d 0 bench
dir bench
concurrency 24
panic: pebble: inconsistent batch count

goroutine 1 [running]:
github.com/petermattis/pebble.(*memTable).apply(0xc00012a0f0, 0xc00010f7a8, 0x38f2afc, 0xc0081b5cf0, 0x152)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/mem_table.go:147 +0x2d1
github.com/petermattis/pebble.(*DB).replayWAL(0xc0000fe800, 0xc00010fa58, 0x659b20, 0x7ea320, 0xc00413abf0, 0x10, 0x38f2b01, 0x0, 0x0)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/open.go:266 +0x3bd
github.com/petermattis/pebble.Open(0x7ffd65180606, 0x5, 0xc000108210, 0x0, 0x0, 0x0)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/open.go:158 +0xedd
main.runTest(0x7ffd65180606, 0x5, 0xc00010fc70, 0xc00010fc90, 0xc00010fc80)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/cmd/pebble/test.go:193 +0x2a2
main.runSync(0x7c9800, 0xc0000f60a0, 0x1, 0x5)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/cmd/pebble/sync.go:30 +0xb1
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).execute(0x7c9800, 0xc0000f6050, 0x5, 0x5, 0x7c9800, 0xc0000f6050)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x7c9a60, 0x0, 0x625ce8, 0x16)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).Execute(0x7c9a60, 0x7b2360, 0x6220cc)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
        /home/tschottdorf/go/src/github.com/petermattis/pebble/cmd/pebble/main.go:49 +0x34a
```

I had previously run the same command for around an hour on the same directory. The error first occurred when I terminated the program and ran `<same> | log` by accident, but now the error also appears right away using `<same>`

Repros easily with fresh directory:

```
tschottdorf@gceworker-tschottdorf:~/go/src/github.com/petermattis/pebble$ pebble sync -c 24 -d 0 bench
dir bench
concurrency 24
_elapsed____ops/sec__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
      1s     5452.7      4.5      5.0      5.2      7.9
      2s     5712.9      4.2      5.0      5.2      7.1
^C
_elapsed_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
    2.5s          14081         5605.4      4.3      4.5      5.0      5.5      7.9

tschottdorf@gceworker-tschottdorf:~/go/src/github.com/petermattis/pebble$ pebble sync -c 24 -d 0 bench | x
x: command not found
panic: pebble: inconsistent batch count

goroutine 1 [running]:
github.com/petermattis/pebble.(*memTable).apply(0xc0001120f0, 0xc0000f97a8, 0x124a3, 0xc0081a6fd0, 0x166)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/mem_table.go:147 +0x2d1
github.com/petermattis/pebble.(*DB).replayWAL(0xc0000e8800, 0xc0000f9a58, 0x659b20, 0x7ea320, 0xc00413e070, 0x10, 0x124a8, 0x0, 0x0)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/open.go:266 +0x3bd
github.com/petermattis/pebble.Open(0x7fffb5159606, 0x5, 0xc0000f2210, 0x0, 0x0, 0x0)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/open.go:158 +0xedd
main.runTest(0x7fffb5159606, 0x5, 0xc0000f9c70, 0xc0000f9c90, 0xc0000f9c80)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/cmd/pebble/test.go:193 +0x2a2
main.runSync(0x7c9800, 0xc0000e00a0, 0x1, 0x5)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/cmd/pebble/sync.go:30 +0xb1
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).execute(0x7c9800, 0xc0000e0050, 0x5, 0x5, 0x7c9800, 0xc0000e0050)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x7c9a60, 0x0, 0x625ce8, 0x16)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).Execute(0x7c9a60, 0x7b2360, 0x6220cc)
        /home/tschottdorf/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
        /home/tschottdorf/go/src/github.com/petermattis/pebble/cmd/pebble/main.go:49 +0x34a
```

The batch at the offending entry is inconsistent in that it claims to contain 5 ops (`b.count() == 5`) but it actually only contains three:

```
panic: pebble: inconsistent batch count: last seqNum is 83033, expected 83035
history:
add at seqNum=83030
add at seqNum=83031
add at seqNum=83032
break
```

This seems reminiscient of examples discussed in https://blog.dgraph.io/post/alice/. I'm running with `db.NoSync`, so it looks like I'm just corrupting the WAL. This is expected, right? And couldn't this also happen if I decided to sync, though presumably more rarely?

Looking into this, doing the same repro steps I ended up with a different panic message:

```
panic: allocation failed because arena is full

goroutine 1 [running]:
github.com/petermattis/pebble.(*DB).replayWAL(0xc000088c00, 0xc0000e9a58, 0x125d300, 0x13f1e80, 0xc0000a1380, 0x10, 0x706412, 0x0, 0x0)
        /Users/justin/go/src/github.com/petermattis/pebble/open.go:258 +0xa70
github.com/petermattis/pebble.Open(0x7ffeefbff8dc, 0x5, 0xc0000e0210, 0x0, 0x0, 0x0)
        /Users/justin/go/src/github.com/petermattis/pebble/open.go:158 +0xedd
main.runTest(0x7ffeefbff8dc, 0x5, 0xc0000e9c70, 0xc0000e9c90, 0xc0000e9c80)
        /Users/justin/go/src/github.com/petermattis/pebble/cmd/pebble/test.go:193 +0x2a2
main.runSync(0x13d17a0, 0xc0000d00a0, 0x1, 0x5)
        /Users/justin/go/src/github.com/petermattis/pebble/cmd/pebble/sync.go:30 +0xb1
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).execute(0x13d17a0, 0xc0000d0050, 0x5, 0x5, 0x13d17a0, 0xc0000d0050)
        /Users/justin/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x13d1a00, 0x0, 0x1228dc9, 0x16)
        /Users/justin/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/petermattis/pebble/vendor/github.com/spf13/cobra.(*Command).Execute(0x13d1a00, 0x13ba300, 0x1225335)
        /Users/justin/go/src/github.com/petermattis/pebble/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
        /Users/justin/go/src/github.com/petermattis/pebble/cmd/pebble/main.go:49 +0x34a
```

Presumably related but perhaps different?

--- a/internal/record/log_writer.go
+++ b/internal/record/log_writer.go
@@ -273,7 +273,7 @@ func (w *LogWriter) WriteRecord(p []byte) (int64, error) {
 		return -1, w.err
 	}
 
-	for i := 0; len(p) > 0; i++ {
+	for i := 0; i == 0 || len(p) > 0; i++ {
 		p = w.emitFragment(i, p)
 	}
 
@@ -307,7 +307,7 @@ func (w *LogWriter) emitFragment(n int, p []byte) []byte {
 	binary.LittleEndian.PutUint16(b.buf[i+4:i+6], uint16(r))
 	atomic.StoreInt32(&b.written, j)
 
-	if blockSize-b.written <= headerSize {
+	if blockSize-b.written < headerSize {
 		// There is no room for another fragment in the block, so fill the
 		// remaining bytes with zeros and queue the block for flushing.
 		for i := b.written; i < blockSize; i++ {

--- a/internal/record/record.go
+++ b/internal/record/record.go
@@ -166,6 +166,9 @@ func (r *Reader) nextChunk(wantFirst bool) error {
 					//
 					// Set r.err to be an error so r.Recover actually recovers.
 					r.err = errors.New("pebble/record: block appears to be zeroed")
+					if !r.recovering {
+						return r.err
+					}
 					r.Recover()
 					continue
 				}

--- a/internal/record/record_test.go
+++ b/internal/record/record_test.go
@@ -28,22 +28,21 @@ func big(partial string, n int) string {
 	return strings.Repeat(partial, n/len(partial)+1)[:n]
 }
 
-// TestZeroBlocks tests that reading nothing but all-zero blocks gives io.EOF.
-// This includes decoding an empty stream.
-func TestZeroBlocks(t *testing.T) {
-	for i := 0; i < 3; i++ {
-		r := NewReader(bytes.NewReader(make([]byte, i*blockSize)))
-		if _, err := r.Next(); err != io.EOF {
-			t.Fatalf("%d blocks: got %v, want %v", i, err, io.EOF)
-		}
-	}
+type recordWriter interface {
+	WriteRecord([]byte) (int64, error)
+	Close() error
 }
 
-func testGenerator(t *testing.T, reset func(), gen func() (string, bool)) {
+func testGeneratorWriter(
+	t *testing.T,
+	reset func(),
+	gen func() (string, bool),
+	newWriter func (io.Writer) recordWriter,
+) {
 	buf := new(bytes.Buffer)
 
 	reset()
-	w := NewWriter(buf)
+	w := newWriter(buf)
 	for {
 		s, ok := gen()
 		if !ok {
@@ -79,6 +78,20 @@ func testGenerator(t *testing.T, reset func(), gen func() (string, bool)) {
 	if _, err := r.Next(); err != io.EOF {
 		t.Fatalf("got %v, want %v", err, io.EOF)
 	}
+}
+
+func testGenerator( t *testing.T, reset func(), gen func() (string, bool)) {
+	t.Run("Writer", func (t *testing.T) {
+		testGeneratorWriter(t, reset, gen, func (w io.Writer) recordWriter {
+			return NewWriter(w)
+		})
+	})
+
+	t.Run("LogWriter", func (t *testing.T) {
+		testGeneratorWriter(t, reset, gen, func (w io.Writer) recordWriter {
+			return NewLogWriter(w)
+		})
+	})
 }
 
 func testLiterals(t *testing.T, s []string) {

